### PR TITLE
Dex 489 - Relationship module

### PR DESCRIPTION
**Pull request by colinwkingen**
Model to create a relationship between two individual animals in the database. Relationship can be initialized with the following signature: 
`Relationship(  individual_1_guid,  individual_2_guid, individual_1_role, individual_2_role,)`

Where roles are represented by strings, and guids are added as strings.

API enabled to create, read and delete Relationships. We don't PATCH a relationship, as it would not makes sense to do something like change the Mother of an individual animal.

Something that might change, like the leader of a group of animals is better served with the SocialGroup construct.

You must be a researcher to post a relationship, and have permission to edit the individuals effected.

Post should look like this:
```json
{
        'individual_1_guid': <individual guid string>,
        'individual_2_guid': <individual guid string>,
        'individual_1_role': 'Mother',
        'individual_2_role': 'Calf',
        'type': 'Family',
        'start_date': <datetime string>
        'end_date':  <datetime string>
}
```

Response should look like this:

```json
{
     'start_date': '2021-12-14T00:05:44.242194+00:00',
     'created': '2021-12-14T00:05:44.242143+00:00',
     'type': None,
     'guid': '04569fdb-bddc-45ce-b69a-9cce347f3fe4',
     'updated': '2021-12-14T00:05:44.242160+00:00',
     'individual_members': [
        {
             'individual_role': 'Mother',
             'individual_guid': '497bd929-35b5-4eff-b0b6-5e032505a021'
        },
       {
            'individual_role': 'Calf',
           'individual_guid': '96d09313-2cb0-4aa7-aa19-69ee1b4ac582'
       }],
    'end_date': '2021-12-14T00:05:44.242201+00:00'
}

```

### CODE REVIEW

File: app/modules/individuals/models.py [hunk 1]
```diff
@@ -91,6 +91,8 @@ class Individual(db.Model, FeatherModel):
         order_by='SocialGroupIndividualMembership.individual_guid',
     )
 
+    # there is a backref'd 'relationships' list of RelationshipIndividualMember accessible here
```
Comment (hwindsor) [inline app/modules/individuals/models.py:94]: In which case I find it easier to read if the thing exists in the module

File: app/modules/relationships/models.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        if (
+            individual_1_guid
+            and individual_2_guid
+            and individual_1_role
+            and individual_2_role
+        ):
+
+            from app.modules.individuals.models import Individual
+
+            individual_1 = Individual.query.get(individual_1_guid)
```
Comment (hwindsor) [inline app/modules/relationships/models.py:75]: What happens if either of these return None?

File: app/modules/individuals/models.py [hunk 3]
```diff
@@ -610,9 +612,17 @@ def find_merge_conflicts(self, individuals):
...
-        with db.session.begin():
-            for group in self.social_groups:
-                db.session.delete(group)
+        with db.session.begin(subtransactions=True):
+            while self.social_groups:
+                db.session.delete(self.social_groups.pop())
+            from app.modules.relationships.models import Relationship
+
+            if self.relationships:
+                for relationship_membership in self.relationships:
```
Comment (hwindsor) [inline app/modules/individuals/models.py:621]: This would probably be better implemented as a relationship_membership delete

File: app/modules/relationships/models.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            member_1 = RelationshipIndividualMember(individual_1, individual_1_role)
+            member_2 = RelationshipIndividualMember(individual_2, individual_2_role)
+
+            self.individual_members.append(member_1)
+            self.individual_members.append(member_2)
+        else:
+            raise ValueError('Relationship needs two individuals, each with a role.')
+
+    def has_individual(self, individual_guid):
+        for individual_member in self.individual_members:
```
Comment (hwindsor) [inline app/modules/relationships/models.py:87]: Consider doing this as a list comprehension
https://github.com/WildMeOrg/houston/blob/main/app/modules/social_groups/models.py#L143-L148

File: app/modules/relationships/models.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            log.debug(
+                'PRINTING MEMBERS: '
+                + str(str(individual_member.individual_guid) + ' == ' + individual_guid)
+            )
+            if str(individual_member.individual_guid) == individual_guid:
+                return True
+        return False
+
+    def get_relationship_role_for_individual(self, individual_guid):
+        if self.has_individual(individual_guid):
```
Comment (hwindsor) [inline app/modules/relationships/models.py:97]: This is inefficient. This implementation goes through the  self.individual_members twice, once to find if it's there, once to get the role.
I'd suggest one local _get_membership_for_guid that returned the membership (using list comprehension) that was used by these two methods

File: app/modules/relationships/models.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            for individual_member in self.individual_members:
+                if str(individual_member.individual_guid) == individual_guid:
+                    return individual_member.individual_role
+        return None
+
+    def __repr__(self):
+        return (
+            '<{class_name}('
+            'guid={self.guid}, '
```
Comment (hwindsor) [inline app/modules/relationships/models.py:106]: Make your life easier by also printing out the relationships. Take it from me, debugginng without this is a PITA.

File: app/modules/relationships/resources.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+                    request_in['individual_1_guid'],
+                    request_in['individual_2_guid'],
+                    request_in['individual_1_role'],
+                    request_in['individual_2_role'],
+                )
+                db.session.add(relationship)
+                for member in relationship.individual_members:
+                    db.session.add(member)
+
+            rtn = {
```
Comment (hwindsor) [inline app/modules/relationships/resources.py:101]: This should really be implemented as a schema

File: app/modules/relationships/resources.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            for member in relationship.individual_members:
+                each_member = {
+                    'individual_role': member.individual_role,
+                    'individual_guid': str(member.individual_guid),
+                }
+                rtn['individual_members'].append(each_member)
+            # ut.embed()
+            return rtn
+        else:
+            log.debug('FAILED TO CREATE RELATIONSHIP!')
```
Comment (hwindsor) [inline app/modules/relationships/resources.py:117]: That's not a debug message, that'a a front end fault Audit log but to generate useful data about the reason you need better detection of what the fault is so that you can report it.

File: app/modules/relationships/resources.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
```
Comment (hwindsor) [inline app/modules/relationships/resources.py:118]: Also see the other code about audit logs for creation, what user created it and how long it took to create.

File: app/modules/users/permissions/rules.py [hunk 10]
```diff
@@ -44,6 +44,8 @@
     ('AuditLog', AccessOperation.READ_PRIVILEGED): ['is_admin'],
     ('SocialGroup', AccessOperation.READ): ['is_researcher'],
     ('SocialGroup', AccessOperation.WRITE): ['is_researcher'],
+    ('Relationship', AccessOperation.READ): ['is_active'],
+    ('Relationship', AccessOperation.WRITE): ['is_active'],
```
Comment (hwindsor) [inline app/modules/users/permissions/rules.py:48]: Do we really want any user to be able to create a relationship between any pair of individuals. 
In the social group code, the requestor must be a researcher and must have view permission to both individuals. 
Surely a relationship should requier the same permissions as social group?

File: tests/modules/relationships/resources/utils.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+from tests import utils as test_utils
+
+PATH = '/api/v1/relationships/'
+
+log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+def create_relationship(flask_app_client, user, expected_status_code=200, data_in={}):
+
+    with flask_app_client.login(user, auth_scopes=('relationships:write',)):
```
Comment (hwindsor) [inline tests/modules/relationships/resources/utils.py:17]: consider using the _via_flask utils as they would simplify these utils a lot.

File: app/modules/relationships/models.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    guid = db.Column(
+        db.GUID, default=uuid.uuid4, primary_key=True
+    )  # pylint: disable=invalid-name
+
+    start_date = db.Column(
+        db.DateTime, index=True, default=datetime.utcnow, nullable=True
+    )
+    end_date = db.Column(db.DateTime, index=True, default=datetime.utcnow, nullable=True)
+
+    type = db.Column(db.String, nullable=True)
```
Comment (naknomum) [inline app/modules/relationships/models.py:56]: old-world we also had a boolean for `bidirectional`.  do we need it here?  i cant find evidence of it really being used, other than simply reported as a property/value on the Relationship.   it seems like `type` + which `role` is assigned to which individual gets us the same thing.  (e.g. "bidirectional" type=familial, if role is `mother` and `calf`, it is implied by the types and roles that there is a direction.)

in summary:  i am all for dropping it; but just wanted to ask.

Comment (colinwkingen) [reply app/modules/relationships/models.py:56]: Yeah, I think role assignment should be a more useful way of determining this. Hard to think of a relationship that _isn't_ bidirectional anyway.

File: app/modules/individuals/models.py (hunk 1 above)
Comment (colinwkingen) [reply app/modules/individuals/models.py:94]: Back_populates wasn't working correctly, it's what I used originally. Switching to backref which does not require the thing to be in the module worked, so I added this note.

File: app/modules/relationships/models.py (hunk 2 above)
Comment (colinwkingen) [reply app/modules/relationships/models.py:75]: Good point, these should be validated. Will do that. I favor guids over the object, as it seems we are more likely to have a guid from an API call than a representation of the object in most cases, and if we can avoid the database hit we should.

File: app/modules/relationships/models.py (hunk 4 above)
Comment (colinwkingen) [reply app/modules/relationships/models.py:87]: Not sure what this wins us besides being a bit more pythonic, but will implement.

File: app/modules/relationships/models.py (hunk 6 above)
Comment (colinwkingen) [reply app/modules/relationships/models.py:114]: Lemme see if I can plug these in from the members table.

File: app/modules/relationships/resources.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+                self,
+            )
+        return None
+
+
+def _user_has_write_permission_on_both_individuals(individual_1_guid, individual_2_guid):
+    individual_1 = Individual.query.get(individual_1_guid)
+    individual_2 = Individual.query.get(individual_2_guid)
+    if (
+        individual_1.current_user_has_edit_permission()
```
Comment (naknomum) [inline app/modules/relationships/resources.py:239]: is it ever a worry that `individual_1` or `_2` might be None?  #NPEtrauma

File: migrations/versions/cfd041b17636_.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        ENTER DESCRIPTION HERE
+    """
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('relationship',
+    sa.Column('created', sa.DateTime(), nullable=False),
+    sa.Column('updated', sa.DateTime(), nullable=False),
+    sa.Column('guid', app.extensions.GUID(), nullable=False),
+    sa.Column('start_date', sa.DateTime(), nullable=True),
+    sa.Column('end_date', sa.DateTime(), nullable=True),
+    sa.Column('type', sa.String(), nullable=True),
```
Comment (naknomum) [inline migrations/versions/cfd041b17636_.py:34]: is `type` able to be None/null?  just curious - have no info myself about it.

File: app/modules/relationships/models.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    )  # pylint: disable=invalid-name
+
+    relationship_guid = db.Column(db.GUID, db.ForeignKey('relationship.guid'))
+    relationship = db.relationship(
+        'Relationship', backref=db.backref('individual_members')
+    )
+
+    individual_guid = db.Column(db.GUID, db.ForeignKey('individual.guid'))
+    individual = db.relationship('Individual', backref=db.backref('relationships'))
+
```
Comment (naknomum) [inline app/modules/relationships/models.py:33]: wonder if you might want to throw in a **unique constraint** on `individual_guid` + `relationship_guid` so that we shouldnt get duplicates?  its likely a very rare thing, but its [pretty much a one-liner](https://github.com/WildMeOrg/houston/blob/main/app/modules/names/models.py#L68-L69).

Comment (naknomum) [reply app/modules/relationships/models.py:33]: and if i were clever, i would know how to make a unique constraint _across the member table and relationship table_ to prevent `indivX` and `indivY` from being in **more than one** relationship of type `type`.   but i am _not_ that clever.  :)

**Review by naknomum — APPROVED** [review]
looks good

File: migrations/versions/cfd041b17636_.py (hunk 14 above)
Comment (colinwkingen) [reply migrations/versions/cfd041b17636_.py:34]: I would venture that it should, especially since migration will not have a type for existing ones.
